### ISSUE TITLE
inherit from empty tuple

### ISSUE THREAD

**Issue by Bergmann89** [boostorg/hana#388]
A class inherited from an empty tuple can not be passed to boost::hana::unpack() (or any other hana method that uses unpack).

```
    template<typename... T>
    struct test : hana::tuple<T...> { };

    hana::for_each(test<> { }, [](auto){
        cout << "fuu" << endl;
    });
```

As soon as you have at least one template parameter, everything works as expected.

```
    hana::for_each(test<int> { }, [](auto){
        cout << "fuu" << endl;
    });
```

Possible solution for that issue:

```
    template <>
    struct unpack_impl<tuple_tag> {

        template<typename Xs, typename = void>
        struct helper
        {
            template <typename F>
            static constexpr decltype(auto) apply(Xs&& xs, F&& f)
                { return hana::unpack(static_cast<Xs&&>(xs).storage_, static_cast<F&&>(f)); }
        };

        template<typename Xs>
        struct helper<Xs,   std::enable_if_t<
                                std::is_base_of<
                                    tuple<>,
                                    std::remove_cv_t<
                                        std::remove_reference_t<Xs>
                                    >
                                >::value
                            >>
        {
            template <typename F>
            static constexpr decltype(auto) apply(Xs&&, F&& f)
                { return static_cast<F&&>(f)(); }
        };

        template <typename Xs, typename F>
        static constexpr decltype(auto) apply(Xs&& xs, F&& f) {
            return helper<Xs>::apply(static_cast<Xs&&>(xs), static_cast<F&&>(f));
        }
    };
```

**Comment by ricejasonf** [comment]
I remember having a discussion on this a couple of years ago, and the conclusion was that inheriting from these data structures is not supported. There doesn't appear to be any evidence of this in the documentation though.

It is half working because of the public member `using hana_tag = tuple_tag` makes `tag_of` yield that tag which is incorrect if your type is meant to be different.

IOW `tuple_tag` is meant for the set of `tuple<T...>` only.

**Comment by ldionne** [comment]
Indeed, inheriting from Hana sequences is not allowed and supported. I will make that clearer in the documentation, and I'll see if I can make `hana::tuple` `final`.

**Comment by ldionne** [comment]
This has been fixed in #391
